File: programmers/Level1/solution/_58_달리기_경주/Solution.java
package programmers.Level1.solution._58_달리기_경주;

import java.util.Arrays;
import java.util.HashMap;

public class Solution {

    public static void main(String[] args) {

        String[] players = { "mumu", "soe", "poe", "kai", "mine" };
        String[] callings = { "kai", "kai", "mine", "mine" };
        String[] expectResult = { "mumu", "kai", "mine", "soe", "poe" };

        Solution solution = new Solution();
        String[] result = solution.solution(players, callings);

        if (Arrays.equals(result, expectResult)) {
            System.out.println("Test Passed");
        } else {
            System.out.println("Test Failed");
        }
    }

    public String[] solution(String[] players, String[] callings) {
        HashMap<String, Integer> playerIndexMap = new HashMap<>();

        for (int i = 0; i < players.length; i++) {
            playerIndexMap.put(players[i], i);
        }

        for (String calling : callings) {
            if (playerIndexMap.containsKey(calling)) {
                int playerIndex = playerIndexMap.get(calling);

                if (playerIndex > 0) {
                    String temp = players[playerIndex - 1];
                    players[playerIndex - 1] = players[playerIndex];
                    players[playerIndex] = temp;
                    playerIndexMap.put(players[playerIndex - 1], playerIndex - 1);
                    playerIndexMap.put(players[playerIndex], playerIndex);
                }
            }

        }

        return players;
    }
}

File: programmers/Level1/solution/_58_달리기_경주/Solution.kt
package programmers.Level1.solution._58_달리기_경주

fun main() {
    val players = arrayOf("mumu", "soe", "poe", "kai", "mine")
    val callings = arrayOf("kai", "kai", "mine", "mine")
    val expectResult = arrayOf("mumu", "kai", "mine", "soe", "poe")

    val result = solution(players, callings)

    if (result.contentEquals(expectResult)) {
        println("Test Passed")
    } else {
        println("Test Failed")
    }
}

fun solution(players: Array<String>, callings: Array<String>): Array<String> {
    val playerIndexMap = HashMap<String, Int>()

    for (i in players.indices) {
        playerIndexMap[players[i]] = i
    }

    for (calling in callings) {
        if (playerIndexMap.containsKey(calling)) {
            val playerIndex = playerIndexMap[calling]!!

            if (playerIndex > 0) {
                val temp = players[playerIndex - 1]
                players[playerIndex - 1] = players[playerIndex]
                players[playerIndex] = temp
                playerIndexMap[players[playerIndex - 1]] = playerIndex - 1
                playerIndexMap[players[playerIndex]] = playerIndex
            }
        }
    }

    return players
}
